add default_service_account resource

bendrucker opened the pull request:
### Description

This PR adds a `kubernetes_default_service_account` resource. This follows the lead of [`aws_default_vpc`](https://www.terraform.io/docs/providers/aws/r/default_vpc.html) and other `aws_default_*` resources. It doesn't create any infrastructure, instead setting minimal state (ID + default secret) and then calling out to `Update` so that any new user-specified config is persisted.

It shares most of its functionality/code with the main `kubernetes_service_account` resource. It overrides:

* The `metadata.name` schema, forcing the value to be "default"
* The `Create` method which doesn't actually create any resources, and just finds the default service account token secret and then calls `Update`

I factored out `getServiceAccountDefaultSecret` so that both resources' `Create` methods could use it. I also updated that to use a field selector to filter service account tokens on the server side. That should help performance in namespaces with large numbers of secrets.

### Acceptance tests
- [x] Have you added an acceptance test for the functionality being added?
- [x] Have you run the acceptance tests on this branch? (If so, please include the test log in a gist)

https://gist.github.com/bendrucker/03b3483eefbd1094aa262c6fa0309b56

### References

Closes #302

### Community Note
<!--- Please keep this note for the community --->
* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

joemiller commented:
Would love to see this merged. Any feedback or changes from the hashicorp folks on next steps?

alexsomesan commented:
I tested this manually using the following configuration and got a failure on apply:
```
resource "kubernetes_namespace" "test" {
	metadata {
		name = "test-876"
	}
}

resource "kubernetes_default_service_account" "test" {
  metadata {
	namespace = kubernetes_namespace.test.metadata.0.name
  }

  secret {
    name = kubernetes_secret.one.metadata.0.name
  }

}

resource "kubernetes_secret" "one" {
  metadata {
    name      = "one"
    namespace = kubernetes_namespace.test.metadata.0.name
  }
}
```
The output looks like this:
```
kubernetes_namespace.test: Creating...
kubernetes_namespace.test: Creation complete after 0s [id=test-876]
kubernetes_secret.one: Creating...
kubernetes_secret.one: Creation complete after 0s [id=test-876/one]
kubernetes_default_service_account.test: Creating...
kubernetes_default_service_account.test: Still creating... [10s elapsed]
kubernetes_default_service_account.test: Still creating... [20s elapsed]
kubernetes_default_service_account.test: Still creating... [30s elapsed]

Error: Waiting for default secret of "test-876/default" to appear

  on test.tf line 7, in resource "kubernetes_default_service_account" "test":
   7: resource "kubernetes_default_service_account" "test" {

```
The secret is created correctly. The service account however is missing the secret association. I'm guessing this is due to a race condition during `resourceKubernetesDefaultServiceAccountCreate`  where the update at the end of it gets called too soon and the default service account is not yet there. I'd suggest trying to add retries around `getServiceAccountDefaultSecret` to make sure it retrieves a valid default service account.

TIP: running Terraform with TRACE debug level will show you the API calls being made and their contents. You can follow the order of operations there.
```
TF_LOG=TRACE terraform apply
```

bendrucker commented:
Looking into it...

bendrucker commented:
Thanks for testing @alexsomesan, definitely an oversight on my part. Should be fixed now!

* I broke the acceptance tests into "basic" (labels/annotations) and "secrets" (secrets/image pull secrets)
* I'm now trying trying to get the ServiceAccount before checking for its secret, retrying if a "not found" error is encountered

I was able to reproduce the race with the ServiceAccount controller via the acceptance tests, but found it was a lot more intermittent. Running in a loop until failure eventually triggered the error:

```sh
while true; do
  TF_ACC=1 go test "./kubernetes" -v -run 'TestAccKubernetesDefaultServiceAccount' -timeout 120m -count 1 || break
done
```

```
=== RUN   TestAccKubernetesDefaultServiceAccount_basic
    testing.go:654: Step 0 error: errors during apply:

        Error: serviceaccounts "default" not found

          on /var/folders/7p/z_kw6fc9693fly126tljpml40000gn/T/tf-test719696721/main.tf line 8:
          (source code not available)


--- FAIL: TestAccKubernetesDefaultServiceAccount_basic (6.70s)
```

I broke up the test so that the ServiceAccount can be created immediately after the namespace. If it also has to wait on several secrets, the race is presumably much less likely.

bendrucker commented:
I've also added a similar wait to the `TestAccKubernetesDefaultServiceAccount_importBasic` test. The `Importer` itself doesn't need to handle a race, since importing implies an already existing namespace and SA. But in an acceptance test, I was able to reproduce a race. So now in addition to applying the configuration, the first import test step also performs a check that waits for the default ServiceAccount's existence.

bendrucker commented on kubernetes/resource_kubernetes_default_service_account_test.go:
```diff
@@ -0,0 +1,229 @@
+
+resource "kubernetes_default_service_account" "test" {
+  metadata {
+		namespace = "${kubernetes_namespace.test.id}"
```
The reason I did this is is that `metadata.0.name` is a statically defined (i.e. "known") value, whereas `id` is a value that is not known until after the namespace is created. I've been operating under the assumption that `.id` prevents a race, whereas `.metadata.0.name` would require `depends_on`. Might be wrong on that, but there's no existing test that depends on creating its own namespace:

```
$ grep -r "kubernetes_namespace.*.metadata.0.name" .
./kubernetes/data_source_kubernetes_namespace_test.go:					resource.TestCheckResourceAttr("data.kubernetes_namespace.test", "metadata.0.name", "kube-system"),
./kubernetes/resource_kubernetes_namespace_test.go:					resource.TestCheckResourceAttr("kubernetes_namespace.test", "metadata.0.name", nsName),
./kubernetes/resource_kubernetes_namespace_test.go:					resource.TestCheckResourceAttr("kubernetes_namespace.test", "metadata.0.name", nsName),
./kubernetes/resource_kubernetes_namespace_test.go:					resource.TestCheckResourceAttr("kubernetes_namespace.test", "metadata.0.name", nsName),
./kubernetes/resource_kubernetes_namespace_test.go:					resource.TestCheckResourceAttr("kubernetes_namespace.test", "metadata.0.name", nsName),
./kubernetes/resource_kubernetes_namespace_test.go:					resource.TestCheckResourceAttr("kubernetes_namespace.test", "metadata.0.name", nsName),
./kubernetes/resource_kubernetes_namespace_test.go:					resource.TestMatchResourceAttr("kubernetes_namespace.test", "metadata.0.name", regexp.MustCompile("^"+prefix)),
./kubernetes/resource_kubernetes_namespace_test.go:					resource.TestCheckResourceAttr("kubernetes_namespace.test", "metadata.0.name", nsName),
./kubernetes/resource_kubernetes_namespace_test.go:					resource.TestCheckResourceAttr("kubernetes_namespace.test", "metadata.0.name", nsName),
```

As for interpolation, 👍🏻, I think I opened this before the transition from 0.11 syntax in acceptance tests and examples.

bendrucker commented:
Awesome, thank you! Seems like whatever is causing `Error: Waiting for default secret of "test-876/default" to appear` is preventing `Update` from then running, so the additional secrets are never set. I fixed the more obvious race condition (default SA doesn't exist yet for a new namespace) but it seems like there's an addition bug with associating the default secret. Will look into it—that hasn't been at all reproducible in my acceptance test runs.

`TestAccKubernetesDefaultServiceAccount_secrets` does have coverage against the secrets and image being present, so as long as the apply succeeds I think everything should work.

bendrucker commented:
Ok, can now confirm the failing example above is now working!

```
kubernetes_namespace.test: Creating...
kubernetes_namespace.test: Creation complete after 0s [id=test-876]
kubernetes_secret.one: Creating...
kubernetes_secret.one: Creation complete after 0s [id=test-876/one]
kubernetes_default_service_account.test: Creating...
kubernetes_default_service_account.test: Creation complete after 0s [id=test-876/default]

Apply complete! Resources: 3 added, 0 changed, 0 destroyed.
```

Leaving notes below for each change.

## Wait Error

```
Error: Waiting for default secret of "test-876/default" to appear
```

Turns out this was not related to acceptance tests versus a full end-to-end TF run, but rather a symptom of having exactly 1 secret specified in configuration. The wait logic checks whether the number of actual secrets matches an input:

```go
len(resp.Secrets) == len(config.Secrets)
```

When `config` is first used to to create a ServiceAccount, this is sound logic. A default ServiceAccount was created by a Kubernetes controller, so we instead need to compare against an empty SA object rather than build up an object from the configuration. Otherwise, the above expression is liable to be permanently true (and cause the resource to time out) when the response contains the default SA token secret and the configuration specifies _exactly one_ secret.

I added some logging to help confirm the problem initially and to aid in future debugging.

## `.metadata.0.name` vs. `.id`

Converted the latter to the former.

I confirmed that these attributes behave equivalently when it comes to the ordering of a Terraform apply graph. The fact that the former is known and the latter is not only affects the plan output, where referencing `id` will cause the target attribute to be `(known after apply)`. As far as the apply graph is concerned, dependencies are drawn between resources, irrespective of their attributes:

```
2020/08/27 14:40:57 [TRACE] Executing graph transform *terraform.ReferenceTransformer
2020/08/27 14:40:57 [DEBUG] ReferenceTransformer: "provider.kubernetes" references: []
2020/08/27 14:40:57 [DEBUG] ReferenceTransformer: "kubernetes_namespace.test (prepare state)" references: []
2020/08/27 14:40:57 [DEBUG] ReferenceTransformer: "kubernetes_default_service_account.test (prepare state)" references: []
2020/08/27 14:40:57 [DEBUG] ReferenceTransformer: "kubernetes_secret.one (prepare state)" references: []
2020/08/27 14:40:57 [DEBUG] ReferenceTransformer: "kubernetes_namespace.test" references: []
2020/08/27 14:40:57 [DEBUG] ReferenceTransformer: "kubernetes_secret.one" references: [kubernetes_namespace.test (prepare state) kubernetes_namespace.test kubernetes_namespace.test]
2020/08/27 14:40:57 [DEBUG] ReferenceTransformer: "kubernetes_default_service_account.test" references: [kubernetes_namespace.test (prepare state) kubernetes_namespace.test kubernetes_namespace.test kubernetes_secret.one (prepare state) kubernetes_secret.one kubernetes_secret.one]
2020/08/27 14:40:57 [TRACE] Completed graph transform *terraform.ReferenceTransformer with new graph:
  kubernetes_default_service_account.test - *terraform.NodeApplyableResourceInstance
    kubernetes_default_service_account.test (prepare state) - *terraform.NodeApplyableResource
    kubernetes_namespace.test - *terraform.NodeApplyableResourceInstance
    kubernetes_namespace.test (prepare state) - *terraform.NodeApplyableResource
    kubernetes_secret.one - *terraform.NodeApplyableResourceInstance
    kubernetes_secret.one (prepare state) - *terraform.NodeApplyableResource
    provider.kubernetes - *terraform.NodeApplyableProvider
  kubernetes_default_service_account.test (prepare state) - *terraform.NodeApplyableResource
    provider.kubernetes - *terraform.NodeApplyableProvider
  kubernetes_namespace.test - *terraform.NodeApplyableResourceInstance
    kubernetes_namespace.test (prepare state) - *terraform.NodeApplyableResource
    provider.kubernetes - *terraform.NodeApplyableProvider
  kubernetes_namespace.test (prepare state) - *terraform.NodeApplyableResource
    provider.kubernetes - *terraform.NodeApplyableProvider
  kubernetes_secret.one - *terraform.NodeApplyableResourceInstance
    kubernetes_namespace.test - *terraform.NodeApplyableResourceInstance
    kubernetes_namespace.test (prepare state) - *terraform.NodeApplyableResource
    kubernetes_secret.one (prepare state) - *terraform.NodeApplyableResource
    provider.kubernetes - *terraform.NodeApplyableProvider
  kubernetes_secret.one (prepare state) - *terraform.NodeApplyableResource
    provider.kubernetes - *terraform.NodeApplyableProvider
  provider.kubernetes - *terraform.NodeApplyableProvider
  ------
2020/08/27 14:40:57 [TRACE] Executing graph transform *terraform.AttachDependenciesTransformer
2020/08/27 14:40:57 [TRACE] AttachDependenciesTransformer: kubernetes_namespace.test depends on []
2020/08/27 14:40:57 [TRACE] AttachDependenciesTransformer: kubernetes_secret.one depends on [kubernetes_namespace.test]
2020/08/27 14:40:57 [TRACE] AttachDependenciesTransformer: kubernetes_default_service_account.test depends on [kubernetes_namespace.test kubernetes_secret.one]
```

## Formatting

The review suggestions were missing a few deprecated interpolations (`"${foo}"`) so I just let [terrafmt](https://github.com/katbyte/terrafmt) handle it:

```sh
terrafmt upgrade012 ./kubernetes/resource_kubernetes_default_service_account_test.go
```

dak1n1 commented:
Looks good! I ran the acceptance tests and did some local testing. It all works as planned. Thanks for this contribution!
